pkg/scheduler/server/internal/pool/loops/connections/connections.go

```diff
@@ -0,0 +1,198 @@
+	Cron       api.Interface
+	CancelPool context.CancelCauseFunc
+}
+
+// connections is a control loop that creates and manages stream connections,
+// piping trigger requests.
+type connections struct {
+	cron       api.Interface
+	cancelPool context.CancelCauseFunc
+	loop       loop.Interface[loops.Event]
+
+	streams    map[uint64]context.CancelCauseFunc
+	streamIDx  uint64
+	streamPool *store.Namespace
+	wg         sync.WaitGroup
+}
+
+func New(opts Options) loop.Interface[loops.Event] {
+	conns := &connections{
+		streams:    make(map[uint64]context.CancelCauseFunc),
+		cancelPool: opts.CancelPool,
+		cron:       opts.Cron,
+		streamPool: store.New(),
+	}
+
+	conns.loop = loop.New(conns, 1024)
+	return conns.loop
+}
+
+func (c *connections) Handle(ctx context.Context, event loops.Event) error {
+	switch e := event.(type) {
+	case *loops.ConnAdd:
+		return c.handleAdd(ctx, e)
+	case *loops.ConnCloseStream:
+		c.handleCloseStream(e)
+	case *loops.TriggerRequest:
+		c.handleTriggerRequest(e)
+	case *loops.Shutdown:
+		c.handleShutdown()
+	default:
+		return fmt.Errorf("unknown connections event type: %T", e)
+	}
+
+	return nil
+}
+
+// handleAdd adds a connection to the pool for a given namespace/appID.
+func (c *connections) handleAdd(ctx context.Context, add *loops.ConnAdd) error {
+	var prefixes []string
+	var appID *string
+
+	reqNamespace := add.Request.GetNamespace()
+	reqAppID := add.Request.GetAppId()
+
+	// To account for backwards compatibility where older clients did not use
+	// this field, we assume a connected client and implement both app jobs, as
+	// well as actor job types. We can remove this in v1.16
+	ts := add.Request.GetAcceptJobTypes()
+	if len(ts) == 0 || slices.Contains(ts, schedulerv1pb.JobTargetType_JOB_TARGET_TYPE_JOB) {
+		log.Infof("Adding a Sidecar connection to Scheduler for appID: %s/%s.", reqNamespace, reqAppID)
+		appID = &add.Request.AppId
+		prefixes = append(prefixes, "app||"+reqNamespace+"||"+reqAppID+"||")
+	}
+
+	if len(ts) == 0 || slices.Contains(ts, schedulerv1pb.JobTargetType_JOB_TARGET_TYPE_ACTOR_REMINDER) {
+		for _, actorType := range add.Request.GetActorTypes() {
+			log.Infof("Adding a Sidecar connection to Scheduler for actor type: %s/%s.", reqNamespace, actorType)
+			prefixes = append(prefixes, "actorreminder||"+reqNamespace+"||"+actorType+"||")
+		}
+	}
+
+	log.Debugf("Marking deliverable prefixes for Sidecar connection: %s/%s: %v.",
+		add.Request.GetNamespace(), add.Request.GetAppId(), prefixes)
+
+	pcancel, err := c.cron.DeliverablePrefixes(ctx, prefixes...)
+	if err != nil {
+		return err
+	}
+
+	log.Debugf("Added a Sidecar connection to Scheduler for: %s/%s.",
+		add.Request.GetNamespace(), add.Request.GetAppId())
+
+	streamIDx := c.streamIDx
+	c.streamIDx++
+
+	streamLoop := stream.New(stream.Options{
+		IDx:      streamIDx,
+		Channel:  add.Channel,
+		Request:  add.Request,
+		ConnLoop: c.loop,
+	})
+
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		err := streamLoop.Run(ctx)
```

Comment:
Shall we use the runner manager here?

Reply:
Unfortunately, we can't here because we are both dynamically creating stream loops over time, and this function _must_ not block- because the connection loop itself is single threaded.

Reply:
ah ok fine by me 👍🏻